src/node_quic_buffer.h

```diff
@@ -0,0 +1,255 @@
+    done_cb done_,
+    void* user_data_,
+    v8::Local<v8::Object> keep_alive_) :
+    buf(buf_),
+    done(done_),
+    user_data(user_data_) {
+    if (!keep_alive.IsEmpty())
+      keep_alive.Reset(keep_alive_->GetIsolate(), keep_alive_);
+  }
+
+  inline ~quic_buffer_chunk() {
+    CHECK(done_called);
+  }
+
+  void Done(int status) {
+    done_called = true;
+    done(status, user_data);
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackFieldWithSize("buf", buf.len);
+  }
+
+  SET_MEMORY_INFO_NAME(quic_buffer_chunk)
+  SET_SELF_SIZE(quic_buffer_chunk)
+};
+
+class QuicBuffer : public MemoryRetainer {
+ public:
+  inline QuicBuffer() :
+    head_(nullptr),
+    tail_(nullptr),
+    size_(0),
+    length_(0) {}
+
+  inline ~QuicBuffer() {
+    Cancel();  // Cancel the remaining data
+    // Verify that all data has been purged, either through
+    // Consume or Cancel
+    CHECK_EQ(length_, 0);
+  }
+
+  // Transfers the state of this QuicBuffer to another
+  // instance, including ownership of the linked list
+  // of uv_buf_t instances. The destination must be
+  // an empty QuicBuffer.
+  inline void TransferAndReset(QuicBuffer* dest) {
+    CHECK(!dest->root_);
+    CHECK_EQ(0, length_);
+    CHECK_EQ(0, size_);
+    dest->head_ = head_;
+    dest->tail_ = tail_;
+    dest->size_ = size_;
+    dest->length_ = length_;
+    head_ = nullptr;
+    tail_ = nullptr;
+    size_ = nullptr;
+    length_ = nullptr;
+    dest->root_.reset(root_.release());
+  }
+
+  // Push one or more uv_buf_t instances into the buffer.
+  // the done_cb callback will be invoked when the last
+  // uv_buf_t in the bufs array is consumed and popped out
+  // of the internal deque. The user_data is passed in to
+  // the done_cb. The keep_alive allows a reference to a
+  // JS object to be kept around until the final uv_buf_t
+  // is consumed.
+  inline void Push(
+      uv_buf_t* bufs,
+      size_t nbufs,
+      done_cb done = default_quic_buffer_chunk_done,
+      void* user_data = nullptr,
+      v8::Local<v8::Object> keep_alive_ = v8::Local<v8::Object>()) {
+    if (nbufs == 0)
+      return done(0, 0, user_data);
+    size_t n = 0;
+    while (nbufs > 1) {
+      Push(bufs[n]);
+      length_ += bufs[n].len;
+      n++;
+      nbufs--;
+    }
+    length_ += bufs[n].len;
+    Push(bufs[n], done, user_data, keep_alive_);
+  }
+
+  // Consume the given number of bytes within the buffer. If amount is
+  // negative, all buffered bytes are consumed.
+  inline void Consume(ssize_t amount = -1) { Consume(0, amount); }
+
+  // Cancels the remaining bytes within the buffer
+  inline void Cancel() { Consume(UV_ECANCELED, -1); }
+
+  // The total buffered bytes
+  inline uint64_t Length() {
+    return length_;
+  }
+
+  // The total number of buffers
+  inline size_t Size() {
+    return size_;
+  }
+
+  // Drain the remaining buffers into the given vector
+  inline size_t DrainInto(std::vector<uv_buf_t*>* list) {
```

Comment:
With a struct as small as `uv_buf_t` it’s not usually worth using pointer indirection, if optimization is the reason for doing this?

Reply:
Yeah, was still working out how I wanted this bit to work... the short version of the issue is this:

1. Whenever we call SendPacket on the `QuicSocket`, we need to hand it some N-number of uv_buf_t instances via the `QuicSocket`'s `SendWrap` class. Right now, passing those off to the `SendWrap` does a memcpy so that the `QuicBuffer` can be reset and can continue to be used while libuv does it's thing sending the data off to the UDP socket. Note that we only get ack callbacks from ngtcp2 for crypto and stream data packets, we don't get acks for other things (like version negotiation and other control packets). So, in some cases, we create the `SendWrap` with it's copy of the data then reset the `QuicBuffer` immediately, in other cases (specifically during the TLS handshake and within `QuicStream`), we create the `SendWrap` with it's copy of the data and only clear data out of the buffer as it is acknowledged. Currently this is done by maintaining a `std::deque` of `QuicBuffer` instances. This new implementation tries to take the approach of moving to a linked list of `quic_buffer_chunk` instances within a `QuicBuffer`. TLS Handshake would still have it's own `QuicBuffer` at the `QuicSession` level, and each `QuicStream` would still have it's own `QuicBuffer`.... hopefully that makes sense  :-D

So, I'm not 100% on any of this implementation at this point and want to find something that works as efficiently as possible with as little copying as possible, that will allow for retransmissions.

Reply:
Yeah, that sounds like you want to have a `std::vector<uv_buf_t>` and return `uv_buf_t` from `Head()` as well, if that’s possible…